### PR TITLE
PP-369: Build system refactoring and demolition

### PR DESCRIPTION
**Pull request by io7m**
**What's this do?**
This performs the passing-a-kidney-stone level of work required to refactor the build system to use Kotlin build scripts. This ended up leading to an absurd rabbit hole of:

  * Upgrading to newer AndroidX components to fix deeply transitive dependency conflicts
  * Requiring API level 24 due to bugs in the Android build tools
  * Deleting unused code that would have otherwise required refactoring
  * Disabling transitive dependency resolution
  * Turning off Jetifier (improves build times)
  * Using JDK 17 to build the code (although the bytecode is still Java 11). Reduces the resources required to build thanks to JDK 17 performance improvements.

The last point requires some explanation: Gradle makes it next to impossible to manage dependency conflicts, despite all of their documentation claiming the contrary. The only reliable way I could get conflicts to stop happening was to actually temporarily turn off transitive dependency resolution to isolate the worst problems. Upon doing this, it actually turned out to be a much nicer way to work, although there is some up-front pain that has to be endured.

Essentially: On adding a new dependency to the application, we now have to also explicitly import the dependencies of that dependency manually. This has some benefits:

  * Because we have to look at the transitive dependencies, we now actually know what goes into the application. If we want to import a dependency that has a hundred transitive dependencies, we might re-evaluate what we're doing instead of importing such a giant mess into the app.
  * Missing dependencies are now (often) compile-time errors. This is a good thing because it forces modules to be honest about what they depend on. During this refactoring I found many, many missing dependencies in modules; this meant that we could break things because modules were depending directly on things being imported by other modules.
  * Missing dependencies that fail to be compile-time errors are now run-time errors with very clear error messaging.

In short: Previously, dependency conflicts could be introduced quietly and would suddenly become a problem one day when the conflict inevitably caused an incomprehensible build or run-time error. With the settings we have now, these problems are signalled much earlier, and in a much clearer way, at the cost of us having to be much more explicit with dependencies.

As a pleasant addition: On my system, the worst-case build time is now six minutes (from ten), and the best case is thirty seconds (from ten minutes).

**Why are we doing this? (w/ JIRA link if applicable)**
https://ebce-lyrasis.atlassian.net/browse/PP-369
https://ebce-lyrasis.atlassian.net/browse/PP-410

**How should this be tested? / Do these changes have associated tests?**
There is still the possibility of lurking missing dependencies, although I have tried almost all of the common use cases in the app. I'm almost certain that A1QA's automated testing will find new crashes.

**Dependencies for merging? Releasing to production?**
Yes, everything.

**Have you updated the changelog?**
Not sure that's applicable.

**Has the application documentation been updated for these changes?**
Heavily commented build system.

**Did someone actually run this code to verify it works?**
You have no idea how many times I've built and run the app in the last two weeks.

### CODE REVIEW

**Comment by io7m** [comment]
@nunommatos Some properties were renamed as part of these changes. I forgot to mention.

```
org.thepalaceproject.drm.enabled -> org.thepalaceproject.adobeDRM.enabled
```

It's now necessary to define each of these to enable support:

```
org.thepalaceproject.lcp.enabled
org.thepalaceproject.findaway.enabled
org.thepalaceproject.overdrive.enabled
```

The CI build for the main branch enables them all. The PR workflow enables none of them.

**Review by nunommts — CHANGES_REQUESTED** [review]
Looks good. There're just three minor issues I left in the comments

File: simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt [hunk 1]
```diff
@@ -507,7 +516,7 @@ object AccountProvidersJSON {
   private fun toStringMap(objectNode: ObjectNode): Map<String, String> {
     val map = mutableMapOf<String, String>()
     for (key in objectNode.fieldNames()) {
-      map[key.toUpperCase()] = JSONParserUtilities.getString(objectNode, key)
+      map[key.uppercase(Locale.getDefault())] = JSONParserUtilities.getString(objectNode, key)
```
[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt:519]: We should be explicit everywhere. This was an IDE suggestion, I'm not sure why it applied it inconsistently... :thinking:

File: simplified-links/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+dependencies {
+    api(libs.irradia.mime.api)
```
[the remark this answers is not in the document]
Comment (io7m) [reply simplified-links/build.gradle.kts:2]: Nope. I'll fix it.

Even if we use `api`, the disabling of transitive dependency resolution means that it really means `implementation` anyway.

File: simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nunommts) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt:519]: But you think it should always be `Locale.ROOT`? I'm more used to see and use the `Locale.getDefault()`

[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: I think I've fixed them all. Chose to use `Locale.ROOT` everywhere.

[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: Oh, weird, I just saw your comment after I'd already replied to the one before it. :dizzy_face: 

I don't think, in any of the cases, we'd want to use the default locale. Using the default locale is for when you actually want language-specific changes to be reflected. If you were parsing text from a book, for example, and you wanted to uppercase the text, you'd use the text's locale to perform a language-correct uppercase. In our code, we're doing things like uppercasing the values of barcode format strings, so we don't want the user's locale settings to affect those transformations.

[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: Is there anywhere in the codebase that you think `Locale.getDefault()` is specifically applicable?

[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: https://github.com/ThePalaceProject/android-core/blob/8c96ead5819f4992b55202fb69b2ff31d602b1d8/simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/ProfileFeedTask.kt#L157C17-L158C6

I think that would be an example of somewhere where `Locale.getDefault()` _is_ applicable.

[the remark this answers is not in the document]
Comment (io7m) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: https://github.com/ThePalaceProject/android-core/blob/8c96ead5819f4992b55202fb69b2ff31d602b1d8/simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/ProfileFeedTask.kt#L327

And that's likely an example where both `Locale.getDefault()` and `Locale.ROOT` would be wrong. :laughing: 

We'd need to use the locale of the book in question, and I'm not sure we have access to it.

[the remark this answers is not in the document]
Comment (nunommts) [reply simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountProvidersJSON.kt]: Maybe for parsing dates. I think right now we're using the `Locale.getDefault()` three times, two of them on the same file (`LocationFragment.kt`) and the other on the `ConfirmationFragment.kt`

**Comment by io7m** [comment]
Ok, that one failing test is the `CookiesTest` that we already know about.

@ray-lee For whatever reason, the web view under test returns all of the correct cookie values _except_ that they no longer contain a `Domain` attribute. I'm not sure if this indicates a real problem, or if it's just that the test setup is fragile. If we think seeing no `Domain` is fine, I can go in and adjust the test values.

**Comment by ray-lee** [comment]
@io7m The whole cookie handling stuff is fragile, since there weren't any public APIs for it. The real test is if SAML login continues to work. If so, the details of the cookies shouldn't really matter.

**Comment by io7m** [comment]
> @io7m The whole cookie handling stuff is fragile, since there weren't any public APIs for it. The real test is if SAML login continues to work. If so, the details of the cookies shouldn't really matter.

I guess I'll change the test values and then wait for people to complain that it doesn't work. :sweat_smile:

**Review by nunommts — APPROVED** [review]

**Comment by io7m** [comment]
Here goes nothing...

**Pull request merged**